util/progress/progressui/display.go

```diff
@@ -276,6 +286,138 @@ func (d *plainDisplay) done() {
 	d.t.printErrorLogs(d.t.w)
 }
 
+type jsonDisplay struct {
+	t    *trace
+	logs *logsByDigest
+	enc  *json.Encoder
+}
+
+// newJSONDisplay creates a new SolveStatusDisplay that outputs the completed vertex
+// progress indicator for each vertex in JSON format.
+func newJSONDisplay(w io.Writer, opts ...DisplayOpt) Display {
+	enc := json.NewEncoder(w)
+	enc.SetIndent("", "  ")
+	return Display{
+		disp: &jsonDisplay{
+			t: newTrace(w, false),
+			logs: &logsByDigest{
+				byDigest: make(map[digest.Digest]*logMux),
+			},
+			enc: enc,
+		},
+	}
+}
+
+func (d *jsonDisplay) init(displayLimiter *rate.Limiter) {
+	// Display limiter is ignored for the JSON display.
+}
+
+func (d *jsonDisplay) update(ss *client.SolveStatus) {
+	d.t.update(ss, 80)
+
+	// JSON format keeps the logs separated into their streams
+	// and takes the aggregated result. The trace does not do that
+	// so we use and update logMux for log messages by digest/stream.
+	for _, l := range ss.Logs {
+		d.logs.accept(l)
+	}
+
+	// Only look at the digests that have updates.
+	// If a digest doesn't have an update, it can't
+	// have been completed.
+	for dgst := range d.t.updates {
+		v, ok := d.t.byDigest[dgst]
+		if !ok {
+			continue
+		}
+
+		if v.isCompleted() {
+			d.printVtx(v)
+		}
+	}
+}
+
+func (d *jsonDisplay) printVtx(v *vertex) {
+	statuses := make([]VertexStatus, len(v.statuses))
+	for i, status := range v.statuses {
+		statuses[i] = VertexStatus{
+			ID:   status.ID,
+			Name: status.Name,
+			// Original structure has these values suitable for
+			// streaming. In those circumstances, we don't know
+			// the total. But, in this circumstance, we're printing
+			// this after the operation is complete so total = current.
+			Total:     status.Current,
+			Started:   status.Started,
+			Completed: status.Completed,
+		}
+	}
+
+	var warnings []VertexWarning
+	if len(v.warnings) > 0 {
+		warnings = make([]VertexWarning, len(v.warnings))
+		for i, warning := range v.warnings {
+			warnings[i] = VertexWarning{
+				Level:  warning.Level,
+				Short:  string(warning.Short),
+				Detail: byteArrayToStringArray(warning.Detail),
+				URL:    warning.URL,
+			}
+		}
+	}
+
+	var logsByStream map[string][]VertexLog
+	if mux, ok := d.logs.byDigest[v.Digest]; ok {
+		logsByStream = make(map[string][]VertexLog)
+		for stream, entries := range mux.streams {
+			var name string
+			switch stream {
+			case 1:
+				name = "stdout"
+			case 2:
+				name = "stderr"
+			default:
+				name = strconv.Itoa(stream)
+			}
+
+			logs := make([]VertexLog, 0, len(entries))
+			for _, entry := range entries {
+				var dt time.Duration
+				if ival := v.mostRecentInterval(); ival != nil && ival.start != nil {
```

Comment:
The need to do `mostRecentInterval` is somewhat confusing to me. These times should always be relative to `v.Vertex.StartedAt`. Otherwise, user would not be able to understand the meaning as `ival.strart` is never printed. Is it always guaranteed to be the same value?

Reply:
Fixed. This was due to a misunderstanding on my part for what the intervals represented.